src/util/http/websocket/QueryHub.cpp

```diff
@@ -45,7 +41,23 @@ QueryHub::createOrAcquireDistributorInternal(QueryId queryId) {
   };
   auto distributor = std::make_shared<QueryToSocketDistributor>(
       ioContext_, [this, cleanupAction = std::move(cleanupAction)]() mutable {
-        net::dispatch(globalStrand_, std::move(cleanupAction));
+        auto future =
+            net::dispatch(globalStrand_,
+                          std::packaged_task<void()>(std::move(cleanupAction)));
+        // If we blocked until the cleanup action is executed on different
+        // strand, this could theoretically lead to deadlocks when globalStrand_
+        // is on a different execution context. This is currently not the case,
+        // but it might be in the future, resulting in deadlocks at distance
+        // which is bad. Unfortunately destructors can't be coroutines, which
+        // would just allow us to co_await execution without issues. To prevent
+        // deadlocks, we wait at most 1 second (arbitrarily chosen interval)
+        // for the cleanup to complete, otherwise we throw an exception.
+        auto status = future.wait_for(std::chrono::seconds(1));
+        if (status != std::future_status::ready) {
+          AD_THROW(
+              "Cleanup took longer than 1s to execute, avoiding deadlock"
+              "by throwing an exception instead!");
+        }
```

Comment:
We'll quickly have to talk about this. 
You should never throw from a Destructor, as it might kill your program if it happens during exception handling. And throwing the exception doesn't help here at all I think, as you still wait for the destructors of all elements.